sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -180,4 +182,169 @@ object DateTimeUtils {
+    val bytes = s.getBytes
+    var j = 0
+    var digitsMilli = 0
+    while (j < bytes.length) {
+      val b = bytes(j)
+      val parsedValue = b - 48
+      if (parsedValue < 0 || parsedValue > 9) {
+        if (i < 2) {
+          if (b == '-') {
+            segments(i) = currentSegmentValue
+            currentSegmentValue = 0
+            i += 1
+          } else {
+            return null
+          }
+        } else if (i == 2) {
+          if (b == ' ' || b == 'T') {
+            segments(i) = currentSegmentValue
+            currentSegmentValue = 0
+            i += 1
+          } else {
+            return null
+          }
+        } else if (i < 5) {
+          if (b == ':') {
+            segments(i) = currentSegmentValue
+            currentSegmentValue = 0
+            i += 1
+          } else {
+            return null
+          }
+        } else if (i < 7) {
+          if (b == 'Z') {
+            segments(i) = currentSegmentValue
+            currentSegmentValue = 0
+            i += 1
+            timeZone = Some(43)
+          } else if (b == '-' || b == '+') {
+            segments(i) = currentSegmentValue
+            currentSegmentValue = 0
+            i += 1
+            timeZone = Some(b)
+          } else if (b == '.' && i == 5) {
+            segments(i) = currentSegmentValue
+            currentSegmentValue = 0
+            i += 1
+          } else {
+            return null
+          }
+          if (i == 6  && b != '.') {
+            i += 1
+          }
+        } else if (i > 6) {
+          if (b == ':') {
+            segments(i) = currentSegmentValue
+            currentSegmentValue = 0
+            i += 1
+          } else {
+            return null
+          }
+        }
+      } else {
+        if (i == 6) {
+          digitsMilli += 1
+        }
+        currentSegmentValue = currentSegmentValue * 10 + parsedValue
+      }
+      j += 1
+    }
+    if (i > 8) {
+      return null
+    }
+    segments(i) = currentSegmentValue
+
+    // Hive compatibility 2011-05-06 07:08:09.1000 == 2011-05-06 07:08:09.1
+    if (digitsMilli == 4) {
+      segments(6) = segments(6) / 10
+    }
+
+    // 18:3:1.1 is equals to 18:3:1:100
+    if (digitsMilli == 1) {
+      segments(6) = segments(6) * 100
+    } else if (digitsMilli == 2) {
+      segments(6) = segments(6) * 10
+    }
+
+    if (segments(0) < 0 || segments(0) > 9999 || segments(1) < 1 || segments(1) > 12 ||
+        segments(2) < 1 || segments(2) > 31 || segments(3) < 0 || segments(3) > 23 ||
+        segments(4) < 0 || segments(4) > 59 || segments(5) < 0 || segments(5) > 59 ||
+        segments(6) < 0 || segments(6) > 999 || segments(7) < 0 || segments(7) > 14 ||
+        segments(8) < 0 || segments(8) > 59) {
+      return null
+    }
+    val c = if (timeZone.isEmpty) {
+      Calendar.getInstance()
+    } else {
+      Calendar.getInstance(
+        TimeZone.getTimeZone(f"GMT${timeZone.get.toChar}${segments(7)}%02d:${segments(8)}%02d"))
```

Comment:
Yes, if the given timezone is invalid, but should the result be null instead of an error? Or when will it fail?